### PR TITLE
VisionTest OCL GPU backend - Fix a seg fault for OpticalFlowLK gdf test

### PR DESCRIPTION
**Pull request by hansely**
Allocating two extra offset values for image data type to match with the PR #678 .
This fixes the issue #541.

### CODE REVIEW

**Comment by kiritigowda** [comment]
@hansely any update on this PR?

**Comment by hansely** [comment]
@kiritigowda The current PR fixes the seg fault and does not break anything, But according to Rajy it's not a correct fix so I'm currently working on a better fix.

**Comment by hansely** [comment]
Created a separate function for halfscalegaussian function which checks for image size when loading memory.
This fixes a segfault on the issue #541

**Comment by kiritigowda** [comment]
@paveltc can you test this with Issue #541?

**Comment by kiritigowda** [comment]
@hansely can you review the CTS logs on the CI and verify that there are no failures?

**Comment by hansely** [comment]
> @hansely can you review the CTS logs on the CI and verify that there are no failures?

Verified that all the tests are passing for CTS

File: amd_openvx/openvx/ago/ago_haf_gpu_common.cpp [hunk 1]
```diff
@@ -245,4 +245,227 @@ int HafGpu_Load_Local(int WGWidth, int WGHeight, int LMWidth, int LMHeight, int
...
+		"    int loffset = ly * %d + (lx << %d);\n" // LMWidth, dTypeShift
+		"    int goffset = (gy - %d) * gstride + (gx << %d) - %d;\n" // gyoffset, dTypeShift, gxoffset
+		), LMWidth, LMHeight, WGWidth, WGHeight, LMWidth, dTypeShift, gyoffset, dTypeShift, gxoffset);
+	code += item;
+	int LMHeightRemain = LMHeight - WGHeight;
+	int LMRemain = (LMWidthRemain * LMHeight + (LMWidth - LMWidthRemain) * LMHeightRemain) >> dTypeShift;
+	if (dGroups == 1 && LMWidthRemain > 0 && LMHeightRemain > 0 && LMRemain < (WGWidth * WGHeight)) {
+		// sanity check
+		if (LMWidthRemain & ((1 << dTypeShift) - 1)) {
+			agoAddLogEntry(NULL, VX_FAILURE, "ERROR: HafGpu_Load_Local(%dx%d,%dx%d,(%d,%d)): doesn't support LMWidthRemain=%d with %s\n", WGWidth, WGHeight, LMWidth, LMHeight, gxoffset, gyoffset, LMWidthRemain, dType);
```
Comment (rrawther) [inline amd_openvx/openvx/ago/ago_haf_gpu_common.cpp:306]: change the agoAddLogEntry to reflect the correct function name

File: amd_openvx/openvx/ago/ago_haf_gpu_common.cpp [hunk 2]
```diff
@@ -245,4 +245,227 @@ int HafGpu_Load_Local(int WGWidth, int WGHeight, int LMWidth, int LMHeight, int
...
+			"      goffset = (gy - ly + ry - %d) * gstride + ((gx - lx + rx) << %d) + %d;\n" // gyoffset, dTypeShift, (WGWidth << LMdivWGWidthShift) - gxoffset
+			"      doExtraLoad = (ry < %d) ? true : false;\n" // LMHeight
+			"    }\n"
+			"    if (doExtraLoad) {\n")
+			, LMHeight - WGHeight, WGHeight, LMWidth, WGHeight
+			, LMHeight - WGHeight, WGWidth, (dWidthShift < 0) ? "/" : ">>", (dWidthShift < 0) ? dWidth : dWidthShift, (dWidthShift < 0) ? "- ry *" : "&", (dWidthShift < 0) ? dWidth : dWidth - 1
+			, LMWidth, dTypeShift, (WGWidth << LMdivWGWidthShift)
+			, gyoffset, dTypeShift, (WGWidth << LMdivWGWidthShift) - gxoffset, LMHeight);
+		code += item;
+		if (use_vload) {
```
Comment (rrawther) [inline amd_openvx/openvx/ago/ago_haf_gpu_common.cpp:344]: looks like the tab spacing is not uniform. Please make it more consistant

File: amd_openvx/openvx/ago/ago_haf_gpu_common.cpp [hunk 3]
```diff
@@ -245,4 +245,227 @@ int HafGpu_Load_Local(int WGWidth, int WGHeight, int LMWidth, int LMHeight, int
...
+		for (int y = 0; y < LMHeight; y += WGHeight) {
+			if ((LMHeight - y) < WGHeight) {
+				sprintf(item, "    if (ly < %d) {\n", LMHeight - y);
+				code += item;
+			}
+			if (y > 0) {
+				sprintf(item,
+					"    loffset += %d * %d;\n" // WGHeight, LMWidth
+					"    goffset += %d * gstride;\n" // WGHeight
+                    "    int sidx = gbuf + goffset - p1_buf - p1_offset;\n"
```
Comment (rrawther) [inline amd_openvx/openvx/ago/ago_haf_gpu_common.cpp:363]: adjust spacing to see more readable code

Comment (rrawther) [inline amd_openvx/openvx/ago/ago_haf_gpu_common.cpp:363]: where is p1_buf and p1_offset calculated? If it is calculated before, please update comment in the beginning to explain those

File: amd_openvx/openvx/ago/ago_haf_gpu_common.cpp [hunk 4]
```diff
@@ -245,4 +245,227 @@ int HafGpu_Load_Local(int WGWidth, int WGHeight, int LMWidth, int LMHeight, int
...
+			code += item;
+			if (dGroups > 1) {
+				if (y > 0) {
+					code +=
+						"    loffset_t = loffset;\n"
+						"    goffset_t = goffset;\n";
+				}
+				else {
+					code +=
+						"    int loffset_t = loffset;\n"
```
Comment (rrawther) [inline amd_openvx/openvx/ago/ago_haf_gpu_common.cpp:383]: why loffset_t is declared only for one condition?

File: amd_openvx/openvx/ago/ago_haf_gpu_common.cpp [hunk 5]
```diff
@@ -245,4 +245,227 @@ int HafGpu_Load_Local(int WGWidth, int WGHeight, int LMWidth, int LMHeight, int
...
+				}
+				else {
+					sprintf(item,
+						"    ry = id >> %d;\n" // dWidthShift
+						"    rx = id & %d;\n" // dWidth-1
+						, dWidthShift, dWidth - 1);
+				}
+				code += item;
+                sprintf(item, 
+                    "    int sidx2 = gbuf + goffset - p1_buf - p1_offset + ry * gstride + (rx << %d);\n"
```
Comment (rrawther) [inline amd_openvx/openvx/ago/ago_haf_gpu_common.cpp:443]: please explain what is p1_buf and p1_offset and why it is subtracted here

**Review by rrawther — CHANGES_REQUESTED** [review]
Please address the requested changes

File: amd_openvx/openvx/ago/ago_haf_gpu_common.cpp (hunk 4 above)
Comment (hansely) [reply amd_openvx/openvx/ago/ago_haf_gpu_common.cpp:383]: I didn't change this part. This part is same as the original code.

**Comment by hansely** [comment]
> Please address the requested changes

@rrawther  All the required changes have made.

**Comment by kiritigowda** [comment]
@paveltc can you verify if this passes the OpenVX tests?

**Comment by kiritigowda** [comment]
@hansely turn on the test which has been turned off with this PR - https://github.com/GPUOpen-ProfessionalCompute-Libraries/MIVisionX/blob/b861d5f9b85471394e21bf350472143089ebe524/tests/vision_tests/runVisionTests.py#L120

**Review by kiritigowda — APPROVED** [review]

**Pull request merged**